lib/cli/ui/prompt.rb

```diff
@@ -48,37 +48,20 @@ class << self
         # Free form question when the answer can be empty
         #   CLI::UI::Prompt.ask('What is your opinion on this question?', allow_empty: true)
         #
-        # Question with answers
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
-        #
-        #
         def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true)
```

Comment:
Talking with @lugray we can both sympathize with the thought that `ask` is a bit of a big method. This being said, we want `ask` to remain the public API. Sometimes when you're writing a public API the capabilities of a method may become a bit big, so we sacrifice a bit of single responsibility to accommodate what people think a method will do instead.

In this case, I don't think we should deprecate `options` and instead we should write something like this:
```ruby
def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true, &options_block)
  if default && !allow_empty
    raise(ArgumentError, 'Cannot combine default and allow_empty keywords')
  end

  has_options = options || options_block
  if has_options && (default || is_file)
    raise(ArgumentError, 'Cannot combine default or is_file with explicit options')
  end

  return ask_with_options(..., &options_block) if has_options
  ask_free_form(...)
end

private

def ask_with_options(...)
end

def ask_free_form(...)
end
```

With this in place, we can call "ask" anywhere, this means it does what you always expect it will be able to do. This does sacrifice a bit of nice code structure for a nicer public API, which is why we split it up entirely on the inside.

Does this make sense?

Reply:
That definitely makes sense! Thanks